Create conditional-operator-access-syntax-refinement.md (#9777)
dotnet/csharplang: proposals/conditional-operator-access-syntax-refinement.md

Similar to the `>>` production, no characters of any kind (not even whitespace, comments, or preprocessor directives) are allowed between these tokens.

Conversely, when the `?` token is followed by `[` or `.` in a conditional expression context, there **must** be at
least one intervening character (whitespace, comment, or preprocessor directive) between the `?` and the subsequent
token.

### Grammar changes

The grammar productions for null-conditional operators now explicitly require token adjacency:

```
null_conditional_member_access
    : primary_expression '?' '.' identifier type_argument_list?
    // where '?' and '.' must be adjacent
    
null_conditional_element_access
    : primary_expression '?' '[' argument_list ']'
    // where '?' and '[' must be adjacent
```

### Examples

**Valid null-conditional access (no space):**
```csharp
var x = obj?[index];        // Null-conditional indexing
var y = obj?.Member;        // Null-conditional member access
```

**Valid conditional expressions (with space/comment):**
```csharp
var x = condition ? [1, 2, 3] : [4, 5, 6];    // Conditional with collection expressions

> var y = flag ? .StaticMember : expr;          // Conditional with static member access

Reviewer: ```suggestion
var y = flag ? .StaticMember : expr;          // Conditional with static member access (possible future feature)
```

- **Null-conditional member access (`?.`)**: The `?` and `.` tokens must be adjacent
- **Null-conditional indexing (`?[`)**: The `?` and `[` tokens must be adjacent

Similar to the `>>` production, no characters of any kind (not even whitespace, comments, or preprocessor directives) are allowed between these tokens.

Conversely, when the `?` token is followed by `[` or `.` in a conditional expression context, there **must** be at
least one intervening character (whitespace, comment, or preprocessor directive) between the `?` and the subsequent
token.

### Grammar changes

The grammar productions for null-conditional operators now explicitly require token adjacency:

```g4
null_conditional_member_access
    : primary_expression '?' '.' identifier type_argument_list? dependent_access*

null_conditional_element_access
    : primary_expression '?' '[' argument_list ']' dependent_access*
```

With explicit text stating:

> `?` `.` and `?` `[` are made up of the two tokens `?` followed by either `.` or `[`. Unlike other productions in the syntactic grammar, no
characters of any kind (not even whitespace) are allowed between the two tokens in each of these

> productions.

Reviewer: Fixing line breaks causing some parts of this to not be inside the blockquote
```suggestion
> `?` `.` and `?` `[` are made up of the two tokens `?` followed by either `.` or `[`. Unlike other productions in the syntactic grammar, no characters of any kind (not even whitespace) are allowed between the two tokens in each of these productions.
```

  made up of the two tokens `>` and `>=`. Unlike other productions in the syntactic grammar, no
  characters of any kind (not even whitespace) are allowed between the two tokens in each of these
  productions. These productions are treated specially in order to enable the correct handling of
  type_parameter_lists.

This proposal extends the same concept to null-conditional operators.

## Detailed design

### Parsing rules

The following multi-character operators are modified to require that their constituent tokens be adjacent with no intervening characters:

- **Null-conditional member access (`?.`)**: The `?` and `.` tokens must be adjacent
- **Null-conditional indexing (`?[`)**: The `?` and `[` tokens must be adjacent

Similar to the `>>` production, no characters of any kind (not even whitespace, comments, or preprocessor directives) are allowed between these tokens.

Conversely, when the `?` token is followed by `[` or `.` in a conditional expression context, there **must** be at
least one intervening character (whitespace, comment, or preprocessor directive) between the `?` and the subsequent
token.

### Grammar changes

> The grammar productions for null-conditional operators now explicitly require token adjacency:

Author: ```suggestion
The existing grammar productions for null-conditional operators are unchanged:
```

> ```g4

Author: Yes.  I'll clarify.

> ```g4

Reviewer: Correct, this is currently unchanged from draft-v8. I suggested updating this to draft-v8, since this part of v8 is extremely unlikely to change (and v8 itself is getting close to done anyway).

```csharp
var x = obj? [index];       // Error: space between ? and [
var y = obj? .Member;       // Error: space between ? and .
```

## Drawbacks

### Breaking change

This is a breaking change. Code that currently uses whitespace between `?` and `[` for null-conditional indexing will no longer compile.  Similarly, code like `a?[b]:[c]` would break as well.  

Real-world examples from existing codebases include:
- `var aux_data = ctx? ["auxData"] as JObject;`
- `var baseScheme = SchemeManager.GetHardCodedSchemes ()? ["Base"];`

These patterns would need to be updated to remove the whitespace: `ctx?["auxData"]` and `SchemeManager.GetHardCodedSchemes()?["Base"]`.
Similarly,  if `a?[b]:` was encountered, it would need to be updated to have explicit spaces to
preserve parsing behavior.

### Migration burden

Developers will need to update their code when upgrading to the language version that includes this feature.

When the compiler encounters code that would have been valid under the old rules but is now invalid (e.g., `ctx? ["auxData"]`), it should:

1. Attempt to parse using the new rules (requiring adjacency for null-conditional access)

> 2. If this fails, attempt to parse using the old interpretation

Author: sure.  we can say it won't catch every change in parsing.  But it will catch the cases we see and think would normally be hit here.  And i think it's fine to mention that people using normal formatters and normal editing will not hit this.

These patterns would need to be updated to remove the whitespace: `ctx?["auxData"]` and `SchemeManager.GetHardCodedSchemes()?["Base"]`.
Similarly,  if `a?[b]:` was encountered, it would need to be updated to have explicit spaces to
preserve parsing behavior.

Note: Users using standard .Net tooling (like Visual Studio and `dotnet format`) would not be likely to run into this issue as the standard formatting engine already formats code such that it follows the syntax pattern codified here.  

### Migration burden

Developers will need to update their code when upgrading to the language version that includes this feature.

When the compiler encounters code that would have been valid under the old rules but is now invalid (e.g., `ctx? ["auxData"]`), it should:

1. Attempt to parse using the new rules (requiring adjacency for null-conditional access)
2. If this fails, attempt to parse using the old interpretation
3. If the old interpretation succeeds as a null-conditional access, provide a diagnostic suggesting removal of the whitespace:
   - **Error**: "Whitespace is not allowed between '?' and '[' in null-conditional indexing. Remove the whitespace to fix this error."

>    - **Suggested fix**: `ctx?["auxData"]`

Author: ```suggestion
The compiler will be augmented to parse in the following fashion:

- If a `?` is seen after an expression.  Then
    - If the following token is not a `[` then normal conditional-operator parsing occurs.
    - If the following token is a `[` then the `when_true` part of the conditional operator is parsed out.
        - If what follows is not a `:` and `[` was next to the `?` (e.g. `a?[b]`), then a nullable-access expression is parsed instead.
        - If what follows is not a `:` and `[` was not next to the `?` (e.g. `a ? [b]`), then this is an error.  The parser should attempt to reparse this as a conditional-access expression.  If that succeeds, it should report that the `?` and `[` must be directly next to each other.  Otherwise, it should report a normal error that `:` was missing.
        - Otherwise, what followed was a `:`
            - If `?` was not touching the `[`, this is a conditional operation (e.g. `a ? [b] :`), which should then consume the colon and finish parsing the when_false portion of the operation.
            - Otherwise, this is `a?[b]:` The parser should reparse after the `?` as a nullable-access operation. And return that expression upwards.  The `:` can then be consumed by a higher parse operation (like a higher conditional operation parse).
               - If no such higher parse operation exists to consume the `:`, the parser should see if it had a conditional-access operation followed by the colon (e.g. `a?[b] : c`).  It should then report that a space should be placed after the `?`
 ```

```csharp
var x = obj? [index];       // Error: space between ? and [
var y = obj? .Member;       // Error: space between ? and .
```

## Drawbacks

### Breaking change

This is a breaking change. Code that currently uses whitespace between `?` and `[` for null-conditional indexing will no longer compile.  Similarly, code like `a?[b]:[c]` would break as well.  

Real-world examples from existing codebases include:
- `var aux_data = ctx? ["auxData"] as JObject;`
- `var baseScheme = SchemeManager.GetHardCodedSchemes ()? ["Base"];`

These patterns would need to be updated to remove the whitespace: `ctx?["auxData"]` and `SchemeManager.GetHardCodedSchemes()?["Base"]`.
Similarly,  if `a?[b]:` was encountered, it would need to be updated to have explicit spaces to
preserve parsing behavior.

Note: Users using standard .Net tooling (like Visual Studio and `dotnet format`) would not be likely to run into this issue as the standard formatting engine already formats code such that it follows the syntax pattern codified here.  

### Migration burden

Developers will need to update their code when upgrading to the language version that includes this feature.

> The compiler will be augmented to parse in the following fashion:

Author: ```suggestion
The compiler will be augmented to parse in the following fashion.  Note: the augmentations only affect what errors are reported in code that cannot be legally parsed.  Specifically, to help users potentially affected by this breaking change to update their code accordingly. 
```

Given:

```
conditional_expression

>     : null_coalescing_expression

Reviewer 2: It threw me off that `null_coalescing_expression` is used here. It might be helpful to include the following quote from the standard.

> The controlling conditional expression of the `?:` operator ([§12.19](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md#1219-conditional-operator)) follows the same rules as a *boolean_expression*, but for reasons of operator precedence is classified as a *null_coalescing_expression*.

Real-world examples from existing codebases include:
- `var aux_data = ctx? ["auxData"] as JObject;`
- `var baseScheme = SchemeManager.GetHardCodedSchemes ()? ["Base"];`

These patterns would need to be updated to remove the whitespace: `ctx?["auxData"]` and `SchemeManager.GetHardCodedSchemes()?["Base"]`.
Similarly,  if `a?[b]:` was encountered, it would need to be updated to have explicit spaces to
preserve parsing behavior.

Note: Users using standard .Net tooling (like Visual Studio and `dotnet format`) would not be likely to run into this issue as the standard formatting engine already formats code such that it follows the syntax pattern codified here.  

### Migration burden

Developers will need to update their code when upgrading to the language version that includes this feature.

The compiler will be augmented to parse in the following fashion:

Given:

```
conditional_expression
    : null_coalescing_expression
    | null_coalescing_expression '?' expression ':' expression
    ;
```

> - If a `?` is seen after an null_coalescing_expression.  Then

Reviewer: ```suggestion
- If a `?` is seen after a null_coalescing_expression.  Then
```

        - If what follows is not a `:` and `[` was not next to the `?` (e.g. `a ? [b]`), then this is an error.  The parser should attempt to reparse this as a `null_conditional_element_access` expression.  If that succeeds, it should report that the `?` and `[` must be directly next to each other.  Otherwise, it should report a normal error that `:` was missing.
        - Otherwise, what followed was a `:`
            - If `?` was not touching the `[`, this is a `conditional_expression` (e.g. `a ? [b] :`), which should then consume the `:` and finish parsing the second `expression` portion of the `conditional_expression`.
            - Otherwise, this is `a?[b]:` The parser should reparse after the `?` as a `null_conditional_element_access`. And return that expression upwards.  The `:` can then be consumed by a higher parse operation (like a higher `conditional_expression` parse).
               - If no such higher parse operation exists to consume the `:`, the parser should see if it had a `null_conditional_element_access` followed by the colon (e.g. `a?[b] : c`).  It should then report that a space should be placed after the `?`.

> This approach provides clear guidance for migration, also ensuring that the breaking change is manageable.

Author: ```suggestion
Similar rules would be present for `?` followed by `.`, with `[` replaced with `.` in the above and `null_conditional_element_access` replaced with `null_conditional_member_access`.

This approach provides clear guidance for migration, also ensuring that the breaking change is manageable.
```

This proposal extends the same concept to null-conditional operators.

## Detailed design

### Parsing rules

The following multi-character operators are modified to require that their constituent tokens be adjacent with no intervening characters:

- **Null-conditional member access (`?.`)**: The `?` and `.` tokens must be adjacent
- **Null-conditional indexing (`?[`)**: The `?` and `[` tokens must be adjacent

Similar to the `>>` production, no characters of any kind (not even whitespace, comments, or preprocessor directives) are allowed between these tokens.

Conversely, when the `?` token is followed by `[` or `.` in a conditional expression context, there **must** be at
least one intervening character (whitespace, comment, or preprocessor directive) between the `?` and the subsequent
token.

### Grammar changes

The existing grammar productions for null-conditional operators are unchanged:

```g4
null_conditional_member_access
    : primary_expression '?' '.' identifier type_argument_list?
      (null_forgiving_operator? dependent_access)*
    ;

null_conditional_projection_initializer
    : primary_expression '?' '.' identifier type_argument_list?
    ;

>     

Author: ```suggestion

```
